### ISSUE TITLE
Add ra_tls_parse_measurement support for RA-TLS middle level API

### ISSUE THREAD

**Issue by 0400H** [gramineproject/gramine#1195]
### Description of the feature

Provide a `ra_tls_parse_measurement`([mid-level-ra-tls-interface](https://gramine.readthedocs.io/en/stable/attestation.html#mid-level-ra-tls-interface)[[source]](https://github.com/gramineproject/gramine/blob/67b1b9fbb658930b1df80d286988283df252565e/tools/sgx/ra-tls/ra_tls.h#L58-L117)) API in Gramine to reduce the process of `parse measurement`.

Just like:
```
__attribute__ ((visibility("hidden")))
ra_tls_measurement ra_tls_parse_measurement(const char *crt, size_t len);

struct ra_tls_measurement {
    char mr_enclave[32];
    char mr_signer[32];
    uint16_t isv_prod_id;
    uint16_t isv_svn;
};
```

### Why Gramine should implement it?

As we all know, [mid-level-ra-tls-interface](https://gramine.readthedocs.io/en/stable/attestation.html#mid-level-ra-tls-interface)[[source]](https://github.com/gramineproject/gramine/blob/67b1b9fbb658930b1df80d286988283df252565e/tools/sgx/ra-tls/ra_tls.h#L58-L117) provides an easy programmable API for remote attestation.

In the scenario where need to get the MR value, this interface is difficult to use.
We need 7 steps, just like:
```
// 1. define parse MRs callback
int gramine_parse_mr_callback(const char* mr_enclave, const char* mr_signer,
                              const char* isv_prod_id, const char* isv_svn) {
    struct ra_tls_measurement mr;
    memcpy(mr.mr_enclave, mr_enclave, 32);
    memcpy(mr.mr_signer, mr_signer, 32);
    mr.isv_prod_id = *(uint16_t*)isv_prod_id;
    mr.isv_svn = *(uint16_t*)isv_svn;
    // save MRs to cache
    _ctx_.cache.mrs.insert({0, mr});
    return 0;
}

// 2. parse MRs by using gramine_parse_mr_callback
ra_tls_measurement gramine_parse_measurement(const char *crt, size_t len) {
    // 3. need to use lock in multi-threads
    std::lock_guard<std::mutex> lock(_ctx_.mtx);

    
    // 4. need to set parse callback to parse MRs
    auto set_verify_mr_callback =
        reinterpret_cast<void (*)(int (*)(const char *mr_enclave,
                                          const char *mr_signer,
                                          const char *isv_prod_id,
                                          const char *isv_svn))>(
            _ctx_.verify_lib.get_func("ra_tls_set_measurement_callback"));
    (*set_verify_mr_callback)(gramine_parse_mr_callback);

    // 5. call callback
    gramine_verify_cert(crt, len);
    // 6. get MRs from cache
    ra_tls_measurement mr = _ctx_.cache.mrs[0];

    // 7. reset to default callback
    set_verify_mr_callback =
        reinterpret_cast<void (*)(int (*)(const char *mr_enclave,
                                          const char *mr_signer,
                                          const char *isv_prod_id,
                                          const char *isv_svn))>(
            _ctx_.verify_lib.get_func("ra_tls_set_measurement_callback"));
    (*set_verify_mr_callback)(verify_measurement);

    return mr;
}
```

Is it possible to provide `ra_tls_parse_measurement` API of ra-tls-middle-level-API in gramine?

**Comment by kailun-qin** [comment]
Hi @0400H, thanks for your request!

> In the scenario where need to get the MRs

Would you please kindly elaborate a bit more your scenario/use case? Like how will the MRs obtained be used (guessing e.g., to be stored somewhere else rather than to be verified directly against)? Thanks!

**Comment by 0400H** [comment]
> Hi @0400H, thanks for your request!
> 
> > In the scenario where need to get the MRs
> 
> Would you please kindly elaborate a bit more your scenario/use case? Like how will the MRs obtained be used (guessing e.g., to be stored somewhere else rather than to be verified directly against)? Thanks!

API `ra_tls_parse_measurement` will be used while it is necessary to separate the process of verifying quotes from verifying MR.

```
Client -> Verify Quote Module -> Verify MRs Module -> ...
Note:      low level module      high level module                   
```
For example:
https://cczoo.readthedocs.io/en/latest/Solutions/attestation-secret-provision/index.html

**Comment by dimakuv** [comment]
@0400H I don't understand the security of your proposed flow.

How will the RA-TLS certificate be verified with your proposal? I guess what I'm missing is `verify_measurement()` implementation, as you show in your code snippet:

> `    (*set_verify_mr_callback)(verify_measurement);`

It looks like you're a bit abusing RA-TLS: **given a certificate, you wish to extract the SGX measurements from this certificate**? Is this what you want?

I think it's meaningful to add such an API. We could do it. I just want to understand why the need for this arises, in which exact scenario. Looks like this is for **gRPC RA-TLS** implementation?

**Comment by 0400H** [comment]
> @0400H I don't understand the security of your proposed flow.
> 
> How will the RA-TLS certificate be verified with your proposal? I guess what I'm missing is `verify_measurement()` implementation, as you show in your code snippet:
> 
> > `    (*set_verify_mr_callback)(verify_measurement);`
> 
> It looks like you're a bit abusing RA-TLS: **given a certificate, you wish to extract the SGX measurements from this certificate**? Is this what you want?
> 
> I think it's meaningful to add such an API. We could do it. I just want to understand why the need for this arises, in which exact scenario. Looks like this is for **gRPC RA-TLS** implementation?

@dimakuv 
1. Yes, I want to get MRs from Cert and use a separate module to provide complex validation strategies. The current implementation can be implemented through callback(middle level api), but it needs to acquire the global lock and make multiple copies, which will waste too much performance.

2. Verify Quote guarantees `runtime security`, and verify MR guarantees `Code integrity`. Some usage scenarios of gRPC-RA-TLS need to separate the above two parts, just like `Attestation and Secret Provision Service` scenario.

**Comment by dimakuv** [comment]
@0400H Thanks for explanations, looks reasonable to me. We can add such an API (**a new callback that, given a certificate, extracts the SGX measurements from this certificate**).

**Comment by dimakuv** [comment]
@kailun-qin @lejunzhu @billionairiam I think this task will be a good fit for one of you.

**Comment by kailun-qin** [comment]
> a new callback that, given a certificate, extracts the SGX measurements from this certificate

Is this a "callback" (rather than a normal API)?

The proposal looks good to me. @dimakuv I can take this task.

**Comment by dimakuv** [comment]
Yes, I think this should be a callback, that must be registered in RA-TLS.

**Comment by mkow** [comment]
I'm not sure I understand this proposal, why is this a callback? Why functions called "parse_xyz" are actually verifying attestation instead of only parsing the data? Would this be an alternative flow of RA-TLS verification?
To be honest, I'm really confused by this whole issue... Could you summarize up both the old and new flows and mark the differences?

**Comment by 0400H** [comment]
> > a new callback that, given a certificate, extracts the SGX measurements from this certificate
> 
> Is this a "callback" (rather than a normal API)?
> 
> The proposal looks good to me. @dimakuv I can take this task.

@kailun-qin @dimakuv 

Why use callback? Maybe normal api would be better.

      Memory                               .so
        |                                   |
        v                                   v
Cert(quote has been verified) -> ra_tls_parse_measurement -> MRs -> agent_verify_measurement

**Comment by kailun-qin** [comment]
> Why use callback?

@0400H I'm not sure about this part and that was why I raised the question. It seems to me that a normal API should work fine to serve the purpose - "given a certificate, extracts the SGX measurements from this certificate". @dimakuv may have some other considerations here.

> Why functions called "parse_xyz" are actually verifying attestation instead of only parsing the data?

I guess this is just a bad naming of function (in the code snippet) bringing confusions. I think @0400H can explain further here.

> Would this be an alternative flow of RA-TLS verification? Could you summarize up both the old and new flows and mark the differences?

@0400H just gives an example of their flow above. Per my understanding, they'd like to separate the measurement verification part from RA-TLS (in our current flow  - via `ra_tls_set_measurement_callback`) and have it conducted via another entity (they call it "Policy Manager"). To this end, the'll need to extract/parse the measurements first (after quote verification) and provide them to that entity for further measurement verification.

So sth like:
* old flow: quote verification + measurement verification (via `ra_tls_set_measurement_callback`) in RA-TLS
* new flow: quote verification in RA-TLS --> measurent extraction/parsing via RA-TLS provided API --> measurement verification by another entity

**Comment by 0400H** [comment]
> Why functions called "parse_xyz" are actually verifying attestation instead of only parsing the data?

```
      Memory                               .so
        |                                   |
        v                                   v
Cert(quote has been verified) -> ra_tls_parse_measurement -> MRs -> agent_verify_measurement
```

"parse_xyz" are actually parsing the data instead of verifying attestation.
MRs is verified by a user defined function "agent_verify_measurement".

**Comment by dimakuv** [comment]
I was confused too. Sorry for misunderstanding.

Yes, we can implement a trivial API `ra_tls_measurement ra_tls_parse_measurement(const char *crt, size_t len)`. It shouldn't be a callback.

But I have a question: **why do you need this function to be a part of RA-TLS**?

It should be trivial to implement this function yourself:
1. You have the certificate `crt`
2. Look through this `crt` to find the RA-TLS OID
3. Now you found where the SGX quote is located (inside the OID-specific object)
4. Parse the SGX quote
5. Extract the required measurements from this SGX quote

The code will look similar to what we currently have in RA-TLS internals: https://github.com/gramineproject/gramine/blob/361dfff60161550fd9c77164f42712a795410053/tools/sgx/ra-tls/ra_tls_verify_common.c#L216-L230

So, why do you need RA-TLS to provide such func? Why not do it yourself?

**Comment by dimakuv** [comment]
I'm reluctant to add new APIs to RA-TLS without significant need.

We are currently at a point where RA-TLS is just 3 API functions, and I like this minimality.

**Comment by kailun-qin** [comment]
> 1. For developers, this will cause confusion when low level and middle level APIs are mixed in development.

@0400H In fact, there is no such design of "lower level API" in RA-TLS (so actually no confusions). The APIs currently supported are listed below (i.e., they are the ONLY ones that are expected to be used): https://github.com/gramineproject/gramine/blob/361dfff60161550fd9c77164f42712a795410053/tools/sgx/ra-tls/ra_tls.h#L34-L84.

I think what @dimakuv suggested is to have the logic of extracting/parsing the measurements from a given cert ALL implemented in your code base rather than calling directly RA-TLS internal APIs (which are not designed to be used that way). But you can definitely follow/leverage the similar logic and functions in RA-TLS internals.

> 3. For "minimality", although the middle api satisfies completeness, performance cannot be guaranteed.

I don't quite get this point. "minimality" is mostly for security and design considerations - I agree w/ @dimakuv. By following the above approach where you get measurements w/o interacting w/ RA-TLS callbacks (but make your own functions), is there still any performance concern?

> 2. Developer maybe do not know or need to know how to extract MRs from Cert with low level api when using middle level api.
> 4. This will be a common and important requirement in the modular design of RA-TLS.

Whether it's a common requirement, it may be and I'm open for discussions. And if it indeed shows significant and common need, making such APIs is definitely doable.

**Comment by 0400H** [comment]
> 2\. cmp_crt_pk_against_quote_report_data

Parsing MRs from a Cert(quote has been verified).


> @0400H You mention these two APIs that are implemented in RA-TLS (and you don't want to implement them):
> 
> 1. `find_oid()` -- this one is very trivial, it's just parsing the OID ASN.1 value. See the code: https://github.com/gramineproject/gramine/blob/64a87bb57306f1b64ce217be3901968665df4d9c/tools/sgx/ra-tls/ra_tls_verify_common.c#L112-L174
![logo_notext](https://user-images.githubusercontent.com/16373603/223202956-2da304c2-6dc1-471e-9b92-d9c3867019fd.svg)

> 2. `cmp_crt_pk_against_quote_report_data()` -- I thought that in your flow, you already verified the cert, and that the cert contains the corresponding public key hash? If not, then I agree, re-implementing this function won't be trivial.
> 
> Still, something doesn't add up in your requirements. Why would you care about `cmp_crt_pk_against_quote_report_data()` if your original request was: **given a certificate, extracts the SGX measurements from this certificate**? If you only need to extract SGX measurements from the cert, your code would be simply:
> 
> 1. `find_oid(/*input=*/crt->v3_ext.p, /*output=*/quote)`
> 2. `return quote->body->report_body->mrsigner` and so on
> 
> This feels like 20 lines of code in your application.

Thanks for your reply.

In fact, some normal developers do not need to know these low-level APIs.
Just like gramine users don't need to know how to write sgxsdk code.
If `ra_tls_parse_measurement` is implemented in gramine middle api, all users will get benefit from it.

Reference:
https://github.com/gramineproject/gramine/blob/361dfff60161550fd9c77164f42712a795410053/pal/src/host/linux-sgx/sgx_attest.h#L46-L61
https://github.com/gramineproject/gramine/blob/361dfff60161550fd9c77164f42712a795410053/tools/sgx/ra-tls/ra_tls_verify_common.c#L239-L256
https://github.com/gramineproject/gramine/blob/361dfff60161550fd9c77164f42712a795410053/tools/sgx/ra-tls/ra_tls_verify_dcap.c#L98-L200
https://github.com/gramineproject/gramine/blob/361dfff60161550fd9c77164f42712a795410053/tools/sgx/ra-tls/ra_tls_verify_common.c#L212-L219


// Impl
```
#include "sgx_attest.h"

#define OID(N) \
{ 0x06, 0x09, 0x2A, 0x86, 0x48, 0x86, 0xF8, 0x4D, 0x8A, 0x39, (N) }
static const uint8_t g_quote_oid[] = OID(0x06);
static const size_t g_quote_oid_size = sizeof(g_quote_oid);

struct ra_tls_measurement {
    char mr_enclave[32];
    char mr_signer[32];
    uint16_t isv_prod_id;
    uint16_t isv_svn;
};

// parsing MR from a verified Cert
ra_tls_measurement ra_tls_parse_measurement(const char *crt, size_t len) {
    ra_tls_measurement mr;

    mbedtls_x509_crt crt;
    mbedtls_x509_crt_init(&crt);

    int ret = mbedtls_x509_crt_parse(&crt, der_crt, der_crt_size);
    if (ret < 0)
        goto out;

    mr = parse_mr_from_cert(/*unused data=*/NULL, &crt, /*depth=*/0, /*flags=*/NULL);
out:
    mbedtls_x509_crt_free(&crt);
    return mr;
}

int parse_mr_from_cert(void* data, mbedtls_x509_crt* crt, int depth, uint32_t* flags) {
    sgx_quote_t* quote;
    size_t quote_size;
    ra_tls_measurement mr;

    /* extract SGX quote from "quote" OID extension from crt */
    int ret = find_oid(crt->v3_ext.p, crt->v3_ext.len, g_quote_oid, g_quote_oid_size,
                       (uint8_t**)&quote, &quote_size);
    if (ret < 0 or quote_size < sizeof(*quote))
        goto out;

    sgx_quote_body_t* quote_body = &quote->body;
    memcpy(mr.mr_enclave, &quote_body->report_body.mr_enclave, 32);
    memcpy(mr.mr_signer, &quote_body->report_body.mr_signer, 32);
    mr.isv_prod_id = *(uint16_t*)&quote_body->report_body.isv_prod_id;
    mr.isv_svn = *(uint16_t*)&quote_body->report_body.isv_svn;

out:
    return mr;
}
```

**Comment by mkow** [comment]
> In fact, some normal developers do not have to know these low-level APIs and source code.

Only if they go the most common / standard usage path. If you want to do more fancy things you may need to add some custom code, and this design pattern is present also in other parts of Gramine.
So, I think our point here is that we're not really convinced that this is a common enough case that justifies adding a new official API, which we'll need to support for a long time (it's easy to add APIs, but hard to later fix them or remove).
From my side I'm not sure whether we should add this API, I'm not strongly against, but I'm also not convinced that this is not some unusual corner case which only you'll need (and you can easily implement that part on your side, as @dimakuv said).

> _@kailun-qin wrote:_
>
> So sth like:
> 
> * old flow: quote verification + measurement verification (via `ra_tls_set_measurement_callback`) in RA-TLS
> * new flow: quote verification in RA-TLS --> measurent extraction/parsing via RA-TLS provided API --> measurement verification by another entity

Hmm, one thing I didn't get here: Why can't you invoke this "measurement verification by another entity" from the measurement verification callback in the normal flow? Maybe I just missed something in this long issue...

**Comment by dimakuv** [comment]
I agree with everything @mkow said.

Let me add just one comment/reply.

> In fact, some normal developers do not have to know these low-level APIs and source code.
> Just like gramine users don't need to know how to write sgxsdk code.
> If `ra_tls_parse_measurement` is implemented in gramine middle api, all users will get benefit from it.

But **you are not normal developers**? Based on the docs that you shared (https://cczoo.readthedocs.io/en/latest/Solutions/attestation-secret-provision/index.html), you're implementing your own attestation service. You're basically re-implementing (or "enhancing", whatever word you prefer) the RA-TLS codebase. Thus, it feels like you're already writing the code that normal developers wouldn't write.

> Impl:
> ```
> #include "sgx_attest.h"
> #define OID(N) \
> { 0x06, 0x09, 0x2A, 0x86, 0x48, 0x86, 0xF8, 0x4D, 0x8A, 0x39, (N) }
> ...
> ```

Yes, this implementation looks correct. And ~60 lines of code. So would you like to use your implementation that you shared, or you still would prefer a new API in RA-TLS itself?

To reiterate: I'm not against adding the new API to RA-TLS, it just that:
1. You're the only ones who requested such an API,
2. You already implemented it in ~60 lines of C code, so looks like the problem is solved?

**Comment by dimakuv** [comment]
We haven't received any similar requests for such a feature, so I'm closing this inactive thread.
